File: Day232/btToBST.go
package main

import (
	"fmt"
	"sort"
)

type Node struct {
	data  int
	left  *Node
	right *Node
}

func binaryTreeToBST(root *Node) *Node {
	list := make([]int, 0)
	traverse(root, &list)
	sort.Ints(list)
	i := 0 // Reset i before starting the inorder traversal
	return inorder(root, list, &i)
}

func traverse(root *Node, list *[]int) {
	if root == nil {
		return
	}
	*list = append(*list, root.data)
	traverse(root.left, list)
	traverse(root.right, list)
}

func inorder(root *Node, list []int, i *int) *Node {
	if root == nil {
		return nil
	}
	root.left = inorder(root.left, list, i)
	root.data = list[*i]
	*i++
	root.right = inorder(root.right, list, i)
	return root
}

func main() {
	// Create and manipulate the binary tree here
}

File: Day232/btToBST.java
class Solution
{
    // The given root is the root of the Binary Tree
    // Return the root of the generated BST
    public int i = 0;
     Node binaryTreeToBST(Node root)
    {
       // Your code here
       List<Integer> list = new ArrayList<>();
       traverse(root,list);
       Collections.sort(list);
       inorder(root,list);
       return root;
    }
    void traverse(Node root, List<Integer> list){
        if(root == null)
        return;
        list.add(root.data);
        traverse(root.left,list);
        traverse(root.right,list);
    }
    void inorder(Node root, List<Integer> list){
        if(root == null)
        return;
        inorder(root.left,list);
        root.data = list.get(i);
        i++;
        inorder(root.right,list);
    }
}
 